### ISSUE TITLE
Delegate to Mercurial to determine the read/write nature of a request

### ISSUE THREAD

**Issue by mharbison72** [scm-manager/scm-manager#2169]
## Issue description

The Mercurial maintainers would like to turn on the `experimental.httppostargs` config by default (which uses POST to send command arguments to avoid header size limitations, e.g. #1525), but there's concern about breaking things when software assumes POST means write access.  SCM Manager came up in the discussion.

I see that there's a [whitelist of wire commands](https://github.com/scm-manager/scm-manager/blob/cb8c951cb8530eaa78324ee5f21ecc69aacd4e73/scm-plugins/scm-hg-plugin/src/main/java/sonia/scm/web/WireProtocol.java#L50) that are interpreted as read commands, but there are some (mostly maintenance related) problems with this approach:

- The `getlfile` and `statlfile` commands from the bundled `largefiles` extension is missing
- The list of wire commands doesn't cover the LFS endpoint, so downloading an LFS blob requires write access because the first command is a POST
- The experimental v2 wireprotocol isn't supported at all
- Commands added by 3rd party extensions aren't covered.  I hit an issue yesterday where `hg outgoing` with the `topic` extension got a 403 without write access, but disabling it and leaving only the `evolve` extension enabled worked fine without write access.

#### Problem to be solved

It should be possible to keep the determination of **who** has a particular permission within SCM Manager, and allow Mercurial to decide **what** permission is required for a given operation via the existing Mercurial `web` config options.  This would be compatible with previous versions of Mercurial, and cut down on maintenance issues on your side.

#### Proposed solution

I didn't try this because I wanted to see if this would be acceptable before spending a bunch of time on it (and I have no idea how to write the tests).  But it looks like the `PermissionFilter` base class has a [pretty simple access check](https://github.com/scm-manager/scm-manager/blob/cb8c951cb8530eaa78324ee5f21ecc69aacd4e73/scm-core/src/main/java/sonia/scm/web/filter/PermissionFilter.java#L110) after determining if a request is a write request.  If that were visible to `HgCGIServlet` (say it's made `public static`), and `HgPermissionFilterFactory.createDecorator()` [returned the delegate instead of the filter wrapper](https://github.com/scm-manager/scm-manager/blob/cb8c951cb8530eaa78324ee5f21ecc69aacd4e73/scm-plugins/scm-hg-plugin/src/main/java/sonia/scm/web/HgPermissionFilterFactory.java#L54), the request would end up in `HgCGIServlet`.  Instead of issuing the command with the [wildcard config](https://github.com/scm-manager/scm-manager/blob/cb8c951cb8530eaa78324ee5f21ecc69aacd4e73/scm-plugins/scm-hg-plugin/src/main/java/sonia/scm/web/HgCGIServlet.java#L163) to allow everything, it could be changed to something like this:

```
    if(PermissionFilter.hasPermission(repository, true) {
        config(args, "web.allow_push", "*");
    }
    else {
        config(args, "web.deny_push", "*");
    }

    if(PermissionFilter.hasPermission(repository, false) {
        config(args, "web.allow_read", "*");
    } else {
        config(args, "web.deny_read", "*");
    }
```

Maybe it would be better to use `SecurityUtils.getSubject().getPrincipal()` for the `allow` configs, and only use the wildcard for the anonymous user, but this is clearer for illustration purposes.  The `deny` configs aren't really necessary, but would override anything in Mercurial config files on the system, so it's a bit safer.

I see there's some code that also attempts to parse Mercurial wire commands, and while I didn't trace how all that was used, if it is only used for this authentication stuff, then it can probably go away and the wire protocol stuff will just be a pass through to Mercurial.

**Comment by eheimbuch** [comment]
Thanks for the detailed explanation. We were kinda aware that the current implemenation could be improved since it only supported extensions which we explictly allowed (like the evolve extension). We will discuss this matter and will most likely change this soon.

It seems like you are an hg expert. Is there any more we could or should provide to improve the hg experience with SCM-Manager?

**Comment by mharbison72** [comment]
I'm one of the Mercurial maintainers, and also use SCM Manager for hosting at work, so feel free to ping me if you have questions about anything hg related, even outside of this feature.

I have some general things that I ran into ~6 months ago when testing an upgrade (mostly around LFS support), but I'm still stuck on 1.60 for now because the [subrepo plugin wasn't ported to 2.x](https://groups.google.com/g/scmmanager/c/HbfT6tgiPWQ/m/3SbshvXfBAAJ) .  As such, I hesitate to try to give you a list until I use it for real.  (I haven't had time to try a new installation, but I didn't see the plugin mentioned in recent release notes, or as a repo here on github, so apologies if I missed it.)  The old repo from bitbucket for the plugin is on the software heritage archive, but I didn't try to make it work with 2.x because there's a UI component that's probably changed a lot more than the SCM side of things, and I have no idea how that works.  The simple 3xx redirect in the old plugin is sufficient, but sharing the subrepo into the nested location would be *much* better, as it would mean the repo is properly formed in the filesystem, and then things like `hg verify` work as expected, and can be run in server maintenance scripts.  I do see that you switched the hooks out of the hgrc file and into an extension loaded on demand.  A big thank you for that, because it makes logging into the server and running commands much easier.

**Comment by eheimbuch** [comment]
Hey Matt, we followed your advise and change the internal hg permission handling for SCM-Manager. 

Since you are about to change the httppostargs to be activated on default we are unsure how to handle this check. Can i always assume it is a read request since it will be denied by mercurial? Otherwise we had to ask mercurial if this is a read or write request?: https://github.com/scm-manager/scm-manager/commit/026ffa18fd3fe8c4f2c945407d74fd05386a0552#diff-7cbbfd0b3b2fde0d2fe192405287db139cd9f015cd5a2f37e1619c5b9b3d84c1R72

Any hints are appreciated
